crates/matrix-sdk/src/account.rs

```diff
@@ -1192,15 +1193,20 @@ impl Account {
 
         let index = recent_emojis.iter().position(|(unicode, _)| unicode == emoji);
 
+        // Truncate to the max allowed size, which will remove any emojis that
+        // haven't been used in a very long time. This will also ease the pressure on
+        // `remove` and `insert` shifting lots of elements in the list
+        recent_emojis.truncate(MAX_RECENT_EMOJI_COUNT);
+
         // Remove the emoji from the list if it was present and get it's `count` value
         let count = if let Some(index) = index { recent_emojis.remove(index).1 } else { uint!(0) };
 
-        // Insert the emoji with the updated count at the start of the list, so it's
+        // Create a new list with the emoji in the first position, so it's
         // considered the most recently used emoji
         recent_emojis.insert(0, (emoji.to_owned(), count + uint!(1)));
 
-        // Then truncate to the max allowed size, which will remove any emojis that
-        // haven't been used in a very long time
+        // If the item was a new one, the list will now be `MAX_RECENT_EMOJI_COUNT` + 1,
+        // so truncate it again (this is a no-op if it already has the right size)
         recent_emojis.truncate(MAX_RECENT_EMOJI_COUNT);
```

Comment:
No need to re-truncate after the operation, with my other comment taken into account.